src/solver.rs

```diff
@@ -202,7 +256,7 @@ pub trait DependencyProvider<P: Package, V: Version> {
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 #[cfg_attr(feature = "serde", serde(transparent))]
 pub struct OfflineDependencyProvider<P: Package, V: Version + Hash> {
-    dependencies: Map<P, Map<V, Map<P, Range<V>>>>,
+    dependencies: Map<P, Map<V, AllowedVersions<P, V>>>,
```

Comment:
If the type was introduced because of this type complexity, I think I'd prefer another type defined here.
Maybe something like `type DependencyStore = Map<P, Map<V, ...>>`

Reply:
This exact change wasn't due to `type_complexity` lint, but because it was a bit hard for me to keep that triple map structure in mind. Using newly introduced type here made it easier for me to understand at a glance.

But that may be just me.